### PR TITLE
Updated Upstream (Bukkit/CraftBukkit/Spigot)

### LINKED ISSUE #10907: Spawn player does not receive the effects update the regain their movement speed
### Expected behavior

respawn player should not have any potion effects

### Observed/Actual behavior

spawn player does not receive the potion effects update the regain their movement speed
and drinking milk doesn't remove slow effects from the player.

### Steps/models to reproduce

player.addPotionEffect(new PotionEffect(PotionEffectType.SLOWNESS, 80, 5, true));
then /kill player 

### Plugin and Datapack List

testplugin

### Paper version

Paper-1.21-4

### Other

https://github.com/PaperMC/Paper/assets/69817107/a2a97edf-cf1e-4d91-b0fd-4922448c6050

### LINKED ISSUE #10999: Teleporting from any other dimension to the overworld fully heals you
### Expected behavior

Your health does not change.

### Observed/Actual behavior

Your health is set to full.

### Steps/models to reproduce

1. Go to hell :-)
2. take damage
3. minecraft:tp to someone or something in the overworld. Also works with essentials /tp, /back, /home and seemingly any other plugin teleport.

### Plugin and Datapack List

image

### Paper version

```
[16:28:56 INFO]: This server is running Paper version 1.21-37-master@dd49fba (2024-06-24T02:08:49Z) (Implementing API version 1.21-R0.1-SNAPSHOT)
You are 1 version(s) behind
Download the new version at: https://papermc.io/downloads/paper
Previous version: 1.21-15-2df432f (MC: 1.21)
```

### Other

clean server with new world, defaults on everything.

### PR DESCRIPTION
**Pull request by lynxplay**
Upstream has released updates that appear to apply and compile correctly. This update has not been tested by PaperMC and as with ANY update, please do your own testing

Bukkit Changes:
e86f4dc4 PR-1041: Improve getPlayer(String) docs to clarify it matches the name
9738f005 Fix spawner API documentation
69ebd9fd PR-1034: Add TrialSpawnerSpawnEvent
23cffd9c PR-973: Improve spawner API and add API for Trial Spawners
8bf19163 PR-1038: Clarify HumanEntity#openInventory(InventoryView) JavaDoc
1ff76351 SPIGOT-7732, SPIGOT-7786: Add freezing damage modifier
02161cb4 PR-1034: Add CreatureSpawnEvent.SpawnReason#TRIAL_SPAWNER
f9cb6f34 SPIGOT-7777: All entity potion effects are removed on death
25d548eb PR-1031: Expose Creeper igniter
ccbf0915 SPIGOT-7770: Reserve spaces in shaped recipes for blank ingredients
17f7097c Clarify ambiguity around what is API
71714f0c Remove note from InventoryView JavaDoc
aaf49731 PR-1030: Deprecate more unused methods in UnsafeValues
3a9dc689 SPIGOT-7771: Material.getDefaultAttributes always returns an empty map

CraftBukkit Changes:
c3ceeb6f7 SPIGOT-7814: Call PlayerShearEntityEvent for Bogged
97b1e4f58 Fix wolf armor not dropping from use of shears
fd2ef563a SPIGOT-7812: Revert "SPIGOT-7809: Restore shield/banner conversion for base colours"
f672c351b SPIGOT-7811: Enchantments are applied on sweeping attack even if damage event is cancelled
cfe29350b SPIGOT-7808: Fix implementation of Enchantment#getName() for bad name return
19335f69e SPIGOT-7809: Restore shield/banner conversion for base colours
ae4f5a0be SPIGOT-7805: Fix jukebox deserialization
62e3b73a4 SPIGOT-7804: Fix written book serialization
aac911d26 SPIGOT-7800, SPIGOT-7801: Keep vanilla behaviour for items dropped on player death 13ece474f PR-1429: Implement TrialSpawnerSpawnEvent
bf13e9113 PR-1354: Improve spawner API and add API for Trial Spawners
515fe49e1 Increase outdated build delay
9cd5a19a0 SPIGOT-7794: Cancelling InventoryItemMoveEvent destroys items
ce40c7b14 SPIGOT-7796: Kickplayer newlines not working
5167256ff SPIGOT-7795: Fix damage/stats ignore the invulnerable damage time
f993563ee Improve cross-world teleportation handling
ab29122cf PR-1433: HumanEntity#openInventory(InventoryView) should only support views belonging to the player
764a541c5 SPIGOT-7732: Issue with the "hurt()" method of EntityLiving and invulnerable time
820084b5f SPIGOT-7791: Skull BlockState with null profile causes NullPointerException
5e46f1c82 SPIGOT-7785: Teleporting a player at the right moment can mess up vanilla teleportation
cbd95a6b3 SPIGOT-7772: Include hidden / non-sampled players in player count
3153debc5 SPIGOT-7790: Server crashes after bee nest is forced to update
e77bb26bb SPIGOT-7788: The healing power of friendship advancement is never granted
ee3d7258a SPIGOT-7789: Fix NPE in CraftMetaFirework applyToItem
2889b3a11 PR-1429: Add CreatureSpawnEvent.SpawnReason#TRIAL_SPAWNER
cdd05bc7f SPIGOT-7777: Speed attribute stays after death; missing EntityPotionEffectEvent call
d0e6af2d4 PR-1428: Expose Creeper igniter
d01c70e93 PR-1425: Fix bytecode transformation taking care of class-to-interface compatibility.
b2b08f68c SPIGOT-7770: Fix certain shaped recipes not registering
3f8e4161f PR-1426: Deprecate more unused methods in UnsafeValues
2c9dd879e SPIGOT-7771: Material.getDefaultAttributes always returns an empty map

Spigot Changes:
491f3675 Rebuild patches
0a642bd7 Rebuild patches
8897571b Rebuild patches
cb8cf80c Fix newlines in custom restart message
1aabe506 Rebuild patches

### CODE REVIEW

**Comment by lynxplay** [comment]
Currently in draft because
a) did not have time to look over my own diff yet, a SHIT load of stuff was touched.
b) needs extensive testing because upstream, in their infinite wisdom, fucked with the damage system again
c) needs revisiting of some of our patches (see death item vanilla compat for example).

**Comment by lynxplay** [comment]
- [ ] Needs to check in on nms.Player diff for knockback events, general logic there.
- [ ] Spawner API
- [ ] StopAttackingIfTargetInvalid target logic
- [ ] CraftMetaItem's attribute modifier bs, the patch broke completely as they moved a shit load of stuff.
- [ ] Bogged shearing (its scuffed)

File: patches/server/0289-Mob-Spawner-API-Enhancements.patch [hunk 1]
```diff
@@ -69,43 +69,150 @@ index 1d0964a7f544735a0213d5c7832c71f53db139a9..b90127f9f805fdb5bb43a4b8ad2b1049
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartMobSpawner.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartMobSpawner.java
+index 72e34dbfadcebb26a0707ce095b0d270f4d1d97c..66cf668183cfdb36f567729a40d01d996ed507e1 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartMobSpawner.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartMobSpawner.java
+@@ -16,7 +16,7 @@ import org.bukkit.entity.EntitySnapshot;
+ import org.bukkit.entity.EntityType;
+ import org.bukkit.entity.minecart.SpawnerMinecart;
+ 
+-final class CraftMinecartMobSpawner extends CraftMinecart implements SpawnerMinecart {
++final class CraftMinecartMobSpawner extends CraftMinecart implements SpawnerMinecart, org.bukkit.craftbukkit.spawner.PaperSharedSpawnerLogic {
```
Comment (Machine-Maker) [inline patches/server/0289-Mob-Spawner-API-Enhancements.patch]: Paper comment

File: patches/server/0383-Improve-fix-EntityTargetLivingEntityEvent.patch [hunk 2]
```diff
@@ -33,13 +33,12 @@ index 5aeef564cdaabeed88a52635e56073cca3a9d1f1..fe635e46569c67dac1d3581ee930d1bf
...
--                        if (event.getTarget() == null) {
--                            memoryaccessor.erase();
+-                        if (event.getTarget() != null) {
+-                            entityinsentient.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((CraftLivingEntity) event.getTarget()).getHandle());
 -                            return true;
 -                        }
--                        entityliving = ((CraftLivingEntity) event.getTarget()).getHandle();
 +                        // if (event.getTarget() == null) { // Paper - this is wrong, you are skipping the forgetCallback
-+                        //     memoryaccessor.erase();
++                        //     entityinsentient.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((CraftLivingEntity) event.getTarget()).getHandle());
```
Comment (Machine-Maker) [inline patches/server/0383-Improve-fix-EntityTargetLivingEntityEvent.patch]: I think upstream fixed this completely [here](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/e77bb26bb7c920e9106bf4d2da58dc6b450f4d00#nms-patches%2Fnet%2Fminecraft%2Fworld%2Fentity%2Fai%2Fbehavior%2FBehaviorAttackTargetForget.patch). Our diff can probably be reverted then.

File: patches/server/0544-Improve-item-default-attribute-API.patch
Comment (Machine-Maker) [inline patches/server/0544-Improve-item-default-attribute-API.patch:1]: Patch description/name should probably just be "improving" now that its not fixing anything

File: patches/server/0593-Add-critical-damage-API.patch [hunk 3]
```diff
@@ -28,24 +28,24 @@ index c1d121d83591ca1b5bf9d9406c9622b4f24eafef..aee26dd78953ff43306aaa64161f5b9e
...
                      }
  
-@@ -1351,7 +1352,7 @@ public abstract class Player extends LivingEntity {
+@@ -1352,7 +1353,7 @@ public abstract class Player extends LivingEntity {
                                      float f7 = this.getEnchantedDamage(entityliving2, f6, damagesource) * f2;
  
                                      // CraftBukkit start - Only apply knockback if the damage hits
--                                    if (entityliving2.hurt(this.damageSources().playerAttack(this).sweep(), f7)) {
-+                                    if (entityliving2.hurt(this.damageSources().playerAttack(this).sweep().critical(flag2), f7)) { // Paper - add critical damage API
-                                         entityliving2.knockback(0.4000000059604645D, (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.SWEEP_ATTACK); // CraftBukkit // Paper - knockback events
```
Comment (Machine-Maker) [inline patches/server/0593-Add-critical-damage-API.patch:48]: Where did this knockback call go?

File: patches/server/0968-Revert-to-vanilla-handling-of-LivingEntity-actuallyH.patch [hunk 4]
```diff
@@ -5,19 +5,19 @@ Subject: [PATCH] Revert to vanilla handling of LivingEntity#actuallyHurt
...
      }
  
-     // CraftBukkit start
--    protected boolean actuallyHurt(final DamageSource damagesource, float f) { // void -> boolean, add final
-+    protected boolean actuallyHurt(final DamageSource damagesource, float f) { // void -> boolean, add final // Paper - return false ONLY if event cancelled
-        if (!this.isInvulnerableTo(damagesource)) {
-             final boolean human = this instanceof net.minecraft.world.entity.player.Player;
-             float originalDamage = f;
-@@ -2382,12 +2382,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
+-    protected boolean actuallyHurt(final DamageSource damagesource, float f, final EntityDamageEvent event) { // void -> boolean, add final
```
Comment (Machine-Maker) [inline patches/server/0968-Revert-to-vanilla-handling-of-LivingEntity-actuallyH.patch]: We are just adding a comment here now?

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/server/0918-Restore-vanilla-entity-drops-behavior.patch [hunk 5]
```diff
@@ -31,24 +31,13 @@ index 184f3c7a1304f4f9d2aaeae27172be9d853c30de..3cbb59df34156479d24a8251f2b3acbb
          if (this.shouldDropLoot() && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Paper - fix player loottables running when mob loot gamerule is false
          // SPIGOT-5071: manually add player loot tables (SPIGOT-5195 - ignores keepInventory rule)
          this.dropFromLootTable(damageSource, this.lastHurtByPlayerTime > 0);
--        for (org.bukkit.inventory.ItemStack item : this.drops) {
--            loot.add(item);
--        }
+-        this.dropCustomDeathLoot(this.serverLevel(), damageSource, flag);
+-
+-        loot.addAll(this.drops);
 +        loot.addAll(this.drops); // Paper
```
Comment (Lulu13022002) [inline patches/server/0918-Restore-vanilla-entity-drops-behavior.patch]: The diff is now similar but idk why spigot call dropCustomDeathLoot here

File: patches/server/1009-Optimize-Hoppers.patch [hunk 6]
```diff
@@ -452,7 +454,7 @@ index 92086ca118d55ec49cefa5bf18977f8706e3e4b4..8310d132006043e93c612890514c4c7f
 +
 +                //        itemstack.setCount(j);
 +                //        if (j == 1) {
-+                //            blockEntity.setItem(i, itemstack);
++                //            blockEntity.setItem(i, original);
```
Comment (Lulu13022002) [inline patches/server/1009-Optimize-Hoppers.patch]: This diff is not correctly reflected from the removed code above (change happens in isCancelled check not j == 1)

File: patches/server/1009-Optimize-Hoppers.patch [hunk 7]
```diff
@@ -558,13 +562,13 @@ index 92086ca118d55ec49cefa5bf18977f8706e3e4b4..8310d132006043e93c612890514c4c7f
 +
 +        //    itemstack.setCount(j);
 +        //    if (j == 1) {
-+        //        iinventory.setItem(i, itemstack);
++        //        iinventory.setItem(i, original);
```
Comment (Lulu13022002) [inline patches/server/1009-Optimize-Hoppers.patch]: This diff is not needed (see above comment)

File: patches/server/0924-Add-drops-to-shear-events.patch [hunk 8]
```diff
@@ -233,33 +233,75 @@ index 2de1a2f666da9db1832907e1651dbff948e37252..5c2ed3c39c8eb850f3be1e2ea5b5a7ea
...
+ 
+     @Override
+     public void shear(SoundSource shearedSoundCategory) {
++    // Paper start - shear drop API
++        this.shear(shearedSoundCategory, generateDefaultDrops());
++    }
++
++    @Override
++    public void shear(SoundSource shearedSoundCategory, java.util.List<net.minecraft.world.item.ItemStack> drops) {
++    // Paper end - shear drop API
```
Comment (Lulu13022002) [inline patches/server/0924-Add-drops-to-shear-events.patch:273]: Paper comments should be indented at the level of the code

File: patches/server/0924-Add-drops-to-shear-events.patch [hunk 9]
```diff
@@ -233,33 +233,75 @@ index 2de1a2f666da9db1832907e1651dbff948e37252..5c2ed3c39c8eb850f3be1e2ea5b5a7ea
...
+             // CraftBukkit start
+-            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemstack, hand)) {
+-                this.getEntityData().markDirty(Bogged.DATA_SHEARED); // CraftBukkit - mark dirty to restore sheared state to clients
+-                return InteractionResult.PASS;
++            // Paper start - expose drops in event
++            java.util.List<net.minecraft.world.item.ItemStack> drops = generateDefaultDrops();
++            final org.bukkit.event.player.PlayerShearEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemstack, hand, drops);
++            if (event != null) {
++                if (event.isCancelled()) {
++                    this.getEntityData().markDirty(Bogged.DATA_SHEARED); // CraftBukkit - mark dirty to restore sheared state to clients
```
Comment (Lulu13022002) [inline patches/server/0924-Add-drops-to-shear-events.patch]: using Entity#resendPossiblyDesyncedEntityData would be better

File: patches/server/0548-Add-PlayerKickEvent-causes.patch [hunk 10]
```diff
@@ -519,15 +519,15 @@ index 868c6f5ee5d67b92ad5834e05f9f1795cca879db..c8a4677db69b0f4111d550ca457b9196
...
 --- a/src/main/java/org/spigotmc/RestartCommand.java
 +++ b/src/main/java/org/spigotmc/RestartCommand.java
 @@ -74,7 +74,7 @@ public class RestartCommand extends Command
              // Kick all players
              for ( ServerPlayer p : com.google.common.collect.ImmutableList.copyOf( MinecraftServer.getServer().getPlayerList().players ) )
              {
--                p.connection.disconnect( CraftChatMessage.fromStringOrEmpty( SpigotConfig.restartMessage ) );
-+                p.connection.disconnect( CraftChatMessage.fromStringOrEmpty( SpigotConfig.restartMessage ), org.bukkit.event.player.PlayerKickEvent.Cause.RESTART_COMMAND); // Paper - kick event reason (cause is never used))
+-                p.connection.disconnect( CraftChatMessage.fromStringOrEmpty( SpigotConfig.restartMessage, true ) );
++                p.connection.disconnect( CraftChatMessage.fromStringOrEmpty( SpigotConfig.restartMessage, true ), org.bukkit.event.player.PlayerKickEvent.Cause.RESTART_COMMAND); // Paper - kick event reason (cause is never used))
```
Comment (Lulu13022002) [inline patches/server/0548-Add-PlayerKickEvent-causes.patch:530]: duplicate right bracket at the end but the cause is used?

File: patches/server/0289-Mob-Spawner-API-Enhancements.patch [hunk 11]
```diff
@@ -69,43 +69,150 @@ index 1d0964a7f544735a0213d5c7832c71f53db139a9..b90127f9f805fdb5bb43a4b8ad2b1049
...
-+        net.minecraft.nbt.CompoundTag entity = new net.minecraft.nbt.CompoundTag();
-+        entity.putString("id", net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(net.minecraft.world.entity.EntityType.ITEM).toString());
-+        entity.put("Item", item.save(this.world.getHandle().registryAccess()));
-+        this.getSnapshot().getSpawner().setNextSpawnData(
-+            this.isPlaced() ? this.world.getHandle() : null,
-+            this.getPosition(),
++
++        final net.minecraft.world.item.ItemStack item = CraftItemStack.asNMSCopy(itemStack);
++        final CompoundTag entity = new CompoundTag();
++        entity.putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(EntityType.ITEM).toString());
```
Comment (Lulu13022002) [inline patches/server/0289-Mob-Spawner-API-Enhancements.patch]: can use the constant Entity#ID_TAG

File: patches/server/0968-Revert-to-vanilla-handling-of-LivingEntity-actuallyH.patch (hunk 4 above)
Comment (lynxplay) [reply patches/server/0968-Revert-to-vanilla-handling-of-LivingEntity-actuallyH.patch]: We did too before, removed now tho 👍

File: patches/server/0918-Restore-vanilla-entity-drops-behavior.patch (hunk 5 above)
Comment (lynxplay) [reply patches/server/0918-Restore-vanilla-entity-drops-behavior.patch]: Probably copied from the LivingEntity impl?
Minimized diff.

File: patches/server/0924-Add-drops-to-shear-events.patch (hunk 8 above)
Comment (lynxplay) [reply patches/server/0924-Add-drops-to-shear-events.patch:273]: I do not agree, we are diffing more than the code block here, our diff expands across method declarations.
The comments are hence indented at the level of the code, the method definition.

File: patches/server/0924-Add-drops-to-shear-events.patch (hunk 9 above)
Comment (lynxplay) [reply patches/server/0924-Add-drops-to-shear-events.patch]: It is yea 👍

File: patches/server/0918-Restore-vanilla-entity-drops-behavior.patch (hunk 5 above)
Comment (Doc94) [reply patches/server/0918-Restore-vanilla-entity-drops-behavior.patch]: > Probably copied from the LivingEntity impl? Minimized diff.

Yes copied from the method in LivingEntity where call the two methods related to this.

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**